kani-compiler/Cargo.toml

```diff
@@ -10,6 +10,7 @@ publish = false
 
 [dependencies]
 cbmc = { path = "../cprover_bindings", package = "cprover_bindings", optional = true }
+charon = { git = "https://github.com/AeneasVerif/charon", optional = true, default-features = false, rev = "cdc1dcde447a50cbc20336c79b21b42ac977b7fd" }
```

Comment:
I'm curious. Why not use a git submodule?

Reply:
Good question. I very much prefer using a submodule, and that's how I had it initially, but I couldn't get it to pass CI because of clippy issues in Charon (clippy checks the Charon crate as it is considered part of the Kani workspace). I could not find a way to disable running clippy on a specific crate (the exclude option seems to be broken: https://github.com/rust-lang/rust-clippy/issues/9555), and I eventually gave up.

I'll try to open a PR to address the clippy issues in the library, and once this is done, we can switch to using a git submodule.

Reply:
Happy to fix the clippy issues and run clippy inour CI (as long as you don't have too pedantic settings on). The main issue is that you're using a somewhat old charon commit so you'd have to upgrade to get these fixes

Reply:
Thanks @Nadrieril. I'll try to reproduce the clippy issues I ran into and create a PR that fixes them or at least post them somewhere.

Reply:
I (re-)tried to use a git submodule, and found out that the blocker was not the clippy issues in Charon, but rather that there were compile errors because clippy tries to compile `hax` which is not compatible with our rust toolchain version:
```bash
$ cargo clippy --all  -- -D warnings 
    Checking hax-frontend-exporter v0.1.0-pre.1 (https://github.com/hacspec/hax?branch=main#29d28fcb)
    Checking cprover_bindings v0.55.0 (/home/ubuntu/git/kani/cprover_bindings)
   Compiling kani-compiler v0.55.0 (/home/ubuntu/git/kani/kani-compiler)
    Checking compiletest v0.0.0 (/home/ubuntu/git/kani/tools/compiletest)
    Checking kani_metadata v0.55.0 (/home/ubuntu/git/kani/kani_metadata)
    Checking kani-driver v0.55.0 (/home/ubuntu/git/kani/kani-driver)
error[E0599]: no method named `predicates_defined_on` found for struct `rustc_middle::ty::TyCtxt<'tcx>` in the current scope
   --> /home/ubuntu/.cargo/git/checkouts/hax-310c6deb65d920f9/29d28fc/frontend/exporter/src/state.rs:300:26
    |
300 |             clauses: tcx.predicates_defined_on(did).predicates.sinto(s),
    |                          ^^^^^^^^^^^^^^^^^^^^^
    |
help: there is a method `predicates_of` with a similar name
    |
300 |             clauses: tcx.predicates_of(did).predicates.sinto(s),
    |                          ~~~~~~~~~~~~~

error[E0023]: this pattern has 1 field, but the corresponding tuple variant has 2 fields
    --> /home/ubuntu/.cargo/git/checkouts/hax-310c6deb65d920f9/29d28fc/frontend/exporter/src/types/copied.rs:1290:14
     |
1290 |     Lifetime(Symbol),
     |              ^^^^^^ expected 2 fields, found 1
    --> /home/ubuntu/.rustup/toolchains/nightly-2024-09-12-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_ast/src/token.rs:334:22
     |
     = note: tuple variant has 2 fields
     |
help: use `_` to explicitly ignore each field
     |
1290 |     Lifetime(Symbol, _),
     |                    +++
</snip>
error: could not compile `hax-frontend-exporter` (lib) due to 16 previous errors
```
Apparently, clippy attempts to build the entire Charon package, and not just the library?

Reply:
OK, I got it to work using:
```
cargo clippy --all --exclude charon --no-deps -- -D warnings
```
I'll switch to using a git submodule.

Reply:
Opened https://github.com/AeneasVerif/charon/pull/381 to fix the clippy issues I ran into.

Reply:
Oh, and also ran into errors due to `cargo deny` 😬 

I'll need to update `deny.toml` to work around those.

Reply:
```
$ cargo deny --workspace --all-features check advisories
error[unmaintained]: ansi_term is Unmaintained
  ┌─ /home/ubuntu/git/kani/Cargo.lock:5:1
  │
5 │ ansi_term 0.12.1 registry+https://github.com/rust-lang/crates.io-index
  │ ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ unmaintained advisory detected
  │
  ├ ID: RUSTSEC-2021-0139
  ├ Advisory: https://rustsec.org/advisories/RUSTSEC-2021-0139
  ├ The maintainer has advised that this crate is deprecated and will not receive any maintenance.
    
    The crate does not seem to have much dependencies and may or may not be ok to use as-is.
    
    Last release seems to have been three years ago.
    
    ## Possible Alternative(s)
    
     The below list has not been vetted in any way and may or may not contain alternatives;
    
     - [ansiterm](https://crates.io/crates/ansiterm)
     - [anstyle](https://github.com/epage/anstyle)
     - [console](https://crates.io/crates/console)
     - [nu-ansi-term](https://crates.io/crates/nu-ansi-term)
     - [owo-colors](https://crates.io/crates/owo-colors)
     - [stylish](https://crates.io/crates/stylish)
     - [yansi](https://crates.io/crates/yansi)
    
    ## Dependency Specific Migration(s)
    
     - [structopt, clap2](https://github.com/clap-rs/clap/discussions/4172)
  ├ Announcement: https://github.com/ogham/rust-ansi-term/issues/72
  ├ Solution: No safe upgrade is available!
  ├ ansi_term v0.12.1
    └── assert_tokens_eq v0.1.0
        └── (dev) macros v0.1.0
            └── charon v0.1.36
                └── kani-compiler v0.55.0

error[unmaintained]: difference is unmaintained
   ┌─ /home/ubuntu/git/kani/Cargo.lock:56:1
   │
56 │ difference 2.0.0 registry+https://github.com/rust-lang/crates.io-index
   │ ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ unmaintained advisory detected
   │
   ├ ID: RUSTSEC-2020-0095
   ├ Advisory: https://rustsec.org/advisories/RUSTSEC-2020-0095
   ├ The author of the `difference` crate is unresponsive.
     
     Maintained alternatives:
     
     - [`dissimilar`](https://crates.io/crates/dissimilar)
     
     - [`similar`](https://crates.io/crates/similar)
     
     - [`treediff`](https://crates.io/crates/treediff)
     
     - [`diffus`](https://crates.io/crates/diffus)
   ├ Announcement: https://github.com/johannhof/difference.rs/issues/45
   ├ Solution: No safe upgrade is available!
   ├ difference v2.0.0
     └── assert_tokens_eq v0.1.0
         └── (dev) macros v0.1.0
             └── charon v0.1.36
                 └── kani-compiler v0.55.0

advisories FAILED
```

Reply:
One more: `kani-regression.sh` fails because of unformatted Charon code. I'll need to add an entry in `rustfmt.toml` to ignore `charon`.

Reply:
Sooooooo, to summarize, adding Charon as a git submodule has 3 implications:
- It gets checked by clippy
- It gets checked by cargo fmt
- It gets checked by cargo deny
I worked around all 3 issues, but we should fix the issues upstream and remove the workarounds.

Reply:
The cargo deny issues look easy to fix. Clippy I can add to our CI. Formatting I thought we already had this in our CI, what was the diff?

Reply:
Here's the diff:

https://gist.github.com/zhassan-aws/d72bacfcb7bc93a94fd978d9feb4b71e

Reply:
Having `cargo deny` failing is not a bad thing, is it? I'm actually surprised it wasn't failing before, doesn't it check the entire dependency tree? You should be able to go around `cargo fmt` and `cargo deny`, since this should not be a crate in the workspace.

Reply:
@Nadrieril it looks like the formatting is due to the configuration we're using in https://github.com/model-checking/kani/blob/main/rustfmt.toml

@celinval do you know why we're not using the default formatting configuration?

Reply:
> Having `cargo deny` failing is not a bad thing, is it? I'm actually surprised it wasn't failing before, doesn't it check the entire dependency tree? You should be able to go around `cargo fmt` and `cargo deny`, since this should not be a crate in the workspace.

Good question. It was previously added in the `allow-git` section of the `deny.toml`:
```
allow-git = [
    "https://github.com/AeneasVerif/charon",
]
```
Perhaps this makes `cargo deny` not look at its dependencies?

Reply:
> @celinval do you know why we're not using the default formatting configuration?

Answering my own question: it seems we inherited the non-default configuration from rustc.